### ISSUE TITLE
Assertion/precondition violation using aggregated insert into CGAL::Arrangement_2

### ISSUE THREAD

**Issue by vladaburian** [CGAL/cgal#1272]
Hi, I got a problem inserting a set of segments into an arrangement using aggregated insertion. When looking for the minimal example to post, I have found actually 3 subsets producing 3 different assertion/precondition violations.

I was using the latest release (4.8.1). Hopefuly you can reproduce the problem: [gist](https://gist.github.com/vladaburian/3964f947fa3bdf2c8ba7c68ca1f78640).

FYI I got another issue some time ago when inserting x-monotone polylines which you can find on [stackoverflow](http://stackoverflow.com/questions/34511236/cgal-4-7-arrangement-insertarr-begin-end-crashes). It still fails so far, so maybe both cases are somehow related.

**Comment by sloriot** [comment]
I could reproduce the errors. @efifogel would you have time to have a look before the next release?

**Comment by vladaburian** [comment]
I guess it can be reproduced quite easily but just for completeness those are the errors I got:
```
CGAL ERROR: precondition violation!
Expr: cv.is_in_x_range(p)
File: .../include/CGAL/Arr_segment_traits_2.h
Line: 457

CGAL ERROR: precondition violation!
Expr: nodeP != NULL && nodeP->is_valid()
File: .../include/CGAL/Multiset.h
Line: 339

CGAL ERROR: assertion violation!
Expr: (m_statusLine.size() == 0)
File: .../include/CGAL/Sweep_line_2/Basic_sweep_line_2_impl.h
Line: 229
```
Incremental insertion works well, so it can be used as a workaround as Efi suggested in stackoverflow.

**Comment by efifogel** [comment]
I'm looking at it.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/



On Mon, Jul 18, 2016 at 10:48 AM, Sebastien Loriot <notifications@github.com

**Comment by efifogel** [comment]
I was able to find the problem.
I still don't know how to fix it.
I'm recording the info I got so far.

I's possible to reproduce the problem with only 5 segments:

  segments = {
    {{1,3},{4,0}},
    {{4,0},{5,0}},
    {{0,0},{9,0}},
    {{3,0},{6,0}},
    {{2,0},{8,0}},
  };

Using the case above, the problem occurs when processing the event at (3,0).
At the time when the problem occurs the event at (6,0) has a right curve (2
0 8 0).
At some point due to overlaps during the processing of the event (3,0), an
attempts is made to replace the right curve (2 0 8 0) of the event (6,0)
with (5 0 8 0).
The attempt fails; see the code in Sweep_line_2_impl.h
Sweep_line_2::_add_curve_to_right(). (The failure might be justified---I'm
not sure yet).

Later on, the curve (5 0 8 0) also becomes the left curve of the event
(8,0). It is clipped and becomes (6,0 8,0) (as it should).

When processing the event (6,0) the curve (2 0 8 0) is inserted into the
status line.

When processing the event (8,0), the left curve (6,0 8,0) does not match
the curve on the status line, namely, (2 0 8 0).

Had the replacement above succeeded, the matching would have succeeded as
well.

So, I believe that the culprit is the code in _add_curve_to_right().

A curve that results from an overlap has 2 originators.
When a new curve is about to be inserted to the right curves of an event,
originators, common to the newly inserted curves and existing right curves,
are searched for.
In our case:

Start adding a Curve to the right of (6 0)
New curve:            (5 0 8 0)  [sc1: (0 0 9 0), sc2: (2 0 8 0)]
Existing right curve: (2 0 8 0)  [sc1: (2 0 8 0), sc2: (0 0 9 0)]

The 2 curves have common originators. In this case a recursive call is made
to find common originators of the originators, which results with the
conclusion that, while there is an overlap, the current curve should
remain. This is not surprising, since the two curves have the same
originators (in reverse order).

The new curve, namely (5 0 8 0), is the result of overlapping (0 0 9 0) and
(2 0 8 0). However the overlap is detected while processing the event
(5,0). Therefore, it is is further clipped on the left to (5,0). This
(clipping) information is lost, or at least, it is not recorded through the
originators. Examining the originators is thus insufficient.

That's it for now....


   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
Is that a different bug from the bug that was recently fixed in `master` (#1361)?

**Comment by efifogel** [comment]
It is different.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
Efi, is there any chance that you will find a patch for this bug in the following two weeks?

**Comment by efifogel** [comment]
Chances are slim.
It may require a significant change, and if so, significant testing would
also be needed.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
Ah! I did not know. In that case the changes of regression are significant. Let's schedule the fix for CGAL-4.10, then. Thanks Efi.

**Comment by lrineau** [comment]
Fixed by #1487.
